zulip_bots/zulip_bots/lib.py

```diff
@@ -102,6 +102,45 @@ def __init__(self, name: str, email: str) -> None:
         self.email = email
         self.mention = '@**' + name + '**'
 
+class CommonStorage(Protocol):
+    def put(self, key: Text, value: Any) -> None:
+        ...
+
+    def get(self, key: Text) -> Any:
+        ...
+
+    def contains(self, key: Text) -> bool:
+        ...
+
+class BotHandler(Protocol):
+
+    email: str
+    full_name: str
+    storage: CommonStorage
+
+    def identity(self) -> BotIdentity:
+        ...
+
+    def react(self, message: Dict[str, Any], emoji_name: str) -> Dict[str, Any]:
+        ...
+
+    def send_message(self, message: Dict[str, Any]) -> Dict[str, Any]:
+        ...
+
+    # For ExternalBotHandler and StubBotHandler returns Dict[str, Any]
+    # Whereas EmbeddedBotHandler returns None
+    def send_reply(self, message: Dict[str, Any], response: str, widget_content: Optional[str] = None) -> Any:
```

Comment:
There is also variation in other methods?

Reply:
We should avoid `Any` wherever possible because it’s completely unchecked; mypy will silently allow you to put in one type and get out another. Consider one of

* `-> object` (see [`Any` vs. `object`](https://mypy.readthedocs.io/en/stable/dynamic_typing.html#any-vs-object));
* `-> Optional[Dict[str, Any]]`;
* `-> Dict[str, Any]` and add the return value in all cases;
* `-> None` and remove the return value in all cases.

Reply:
@andersk 
If we use any of:
- `Optional[Dict[str, Any]]`;
- `Dict[str, Any]`;
- `None`

The method will start expecting the same return declaration in the actual class methods.
Tried Overloading as well but didn't work.

Thanks for the object vs Any, comparison. I think I should update it as an object.

Reply:
It’s perfectly [valid](https://mypy-play.net/?mypy=latest&python=3.9&gist=dd4e030f4136c7cd74d074e04fd2c8d8) to implement a protocol method returning `Optional[Dict[str, Any]]` with an implementation method returning `Dict[str, Any]` or `None`, as these are both subtypes of `Optional[Dict[str, Any]]`.

For the other two suggestions, you’ve left off the important part: “and add/remove the return value in all cases”. I think one of these is probably the right plan. Even though all four plans would be accepted by mypy, there’s no actual reason to a return a value in some implementations but not others. What’s the point if the caller can’t rely on it being there?

Avoid overloading; it’s also completely unchecked.

Reply:
But I think we cannot use objects here because it says objects are not indexable, but Dicts are. So, I think Any is the only way to go. The same reason why we can't use `Optional[Dict[str, Any]]`, Optional will give it a possibility of None making it non-indexable as well.

Reply:
`None` doesn’t become any more indexable just because you hid it away in an `Any` type. Mypy won’t be able to find the error, but it’s still an error.

Reply:
(Can you reply with normal replies instead of editing your original comment? It’s confusing to try to have this discussion out of order.)

I don’t see why you’re saying “that is not true for a `Dict` return” when I just demonstrated that it is.

Reply:
I am sorry to edit my message, I thought of condensing everything in one message. Won't do that from now. 
[this](https://mypy-play.net/?mypy=latest&python=3.9&gist=dd4e030f4136c7cd74d074e04fd2c8d8) defines the error.

Reply:
oops it didn't save

Reply:
I am not sure how to share mypy playground:
The error is thrown when we run this:

try this out:
```
from typing import Any, Dict, Optional

from typing_extensions import Protocol

class P(Protocol):
    def foo(self) -> Optional[Dict[str, Any]]:
        ...

class A:
    def foo(self) -> Dict[str, Any]:
        return {"test": "test"}

class B:
    def foo(self) -> None:
        pass

new = A() # type: P
d = new.foo()
print(d["test"])
```

Reply:
Right: `A()` is a valid implementation of `P`, but `d["test"]` is not a valid use of `P`.

If you want `B()` to be a valid implementation, then `d["test"]` cannot be a valid use. If you want `d["test"]` to be a valid use, then `B()` cannot be a valid implementation. One or the other will need to be fixed.

Reply:
Yes, that's my point. I have commented on the reason behind me choosing `Any` is in reply to your next comment.

Reply:
But this is not a reason to choose `Any`!

We are clearly failing to communicate somewhere, so let me try to simplify. Let’s take all the protocols and classes out of this and just look at some basic types.

Mypy rejects this program because it is wrong:

```python
d = None
print(d["test"])
```

Mypy accepts this program, but it is **still just as wrong**:

```python
d: Any = None
print(d["test"])
```

Using `Any` has **not fixed the bug**. It has only silenced mypy.

Our goal is not to silence mypy. Our goal is to fix the bug.

We can fix the bug by having the consumer check for `None`:

```python
d: Optional[Dict[str, str]] = None
if d is not None:
    print(d["test"])
```

or we can fix the bug by having the producer avoid producing `None`:

```python
d: Dict[str, str] = {"test": "yay"}
print(d["test"])
```

Reply:
I thought we shouldn't put `Optional` because `bot_handler` was an instance of `StubBotHandler` which returns `Dict` strictly, but we have made `bot_handler` an instance of `BotHandler` which made the type `Optional[Dict]`.

will fix this error:
```
mypy      | zulip_bots/zulip_bots/bots/incrementor/incrementor.py:37: error: Value of type "object" is not indexable
```
by putting a check and type as Optional[Dict].

I think fix 1:
```python
d: Optional[Dict[str, str]] = None
if d is not None:
    print(d["test"])
```
is better, then we would ideally be able to send any kind of bot_handler to `bot_handler` parameter.

Thanks for helping soo much, and sorry for being a dumb kid.